firedrake/function.py

```diff
@@ -487,3 +500,197 @@ def __idiv__(self, expr):
+        # Called by UFL when evaluating expressions at coordinates
+        return self.at(coord)
+
+    def at(self, arg, *args, **kwargs):
+        """Evaluate function at points."""
+        from mpi4py import MPI
+        comm = MPI.COMM_WORLD
+
+        if args:
+            arg = (arg,) + args
+        arg = np.array(arg, dtype=float)
+
+        dont_raise = kwargs.get('dont_raise', False)
+
+        # Handle f.at(0.3)
+        if not arg.shape:
+            arg = arg.reshape(-1)
+
+        # Immersed not supported
+        tdim = self.function_space().mesh().ufl_cell().topological_dimension()
+        gdim = self.function_space().mesh().ufl_cell().geometric_dimension()
+        if tdim < gdim:
+            raise NotImplementedError("Point is almost certainly not on the manifold.")
+
+        # Validate geometric dimension
+        if arg.shape[-1] == gdim:
+            pass
+        elif len(arg.shape) == 1 and gdim == 1:
+            arg = arg.reshape(-1, 1)
+        else:
+            raise ValueError("Point dimension (%d) does not match geometric dimension (%d)." % (arg.shape[-1], gdim))
+
+        # Check if we have got the same points on each process
+        root_arg = comm.bcast(arg, root=0)
+        same_arg = arg.shape == root_arg.shape and np.allclose(arg, root_arg)
+        diff_arg = comm.allreduce(int(not same_arg), op=MPI.SUM)
+        if diff_arg:
+            raise ValueError("Points to evaluate are inconsistent among processes.")
+
+        def single_eval(x, buf):
+            """Helper function to evaluate at a single point."""
+            err = self._c_evaluate(self.ctypes, x.ctypes.data, buf.ctypes.data)
+            if err == -1:
+                raise PointNotInDomainError(self.function_space().mesh(), x.reshape(-1))
+
+        if not len(arg.shape) <= 2:
+            raise ValueError("Function.at expects point or array of points.")
+        points = arg.reshape(-1, arg.shape[-1])
+        value_shape = self.function_space().ufl_element().value_shape()
+
+        split = self.split()
+        mixed = len(split) != 1
+
+        # Local evaluation
+        l_result = []
+        for i, p in enumerate(points):
+            try:
+                if mixed:
+                    l_result.append((i, tuple(f.at(p) for f in split)))
+                else:
+                    p_result = np.empty(value_shape, dtype=float)
+                    single_eval(points[i:i+1], p_result)
+                    l_result.append((i, p_result))
+            except PointNotInDomainError:
+                # Skip point
+                pass
+
+        # Collecting the results
+        def same_result(a, b):
+            if mixed:
+                for a_, b_ in zip(a, b):
+                    if not np.allclose(a_, b_):
+                        return False
+                return True
+            else:
+                return np.allclose(a, b)
+
+        all_results = comm.gather(l_result, root=0)
```

Comment:
Can do. In that case all processes need to aggregate the results themselves, but that probably isn't much of a problem.